tools/test/build_api/build_api_test.py

```diff
@@ -84,6 +84,59 @@ def test_always_complete_build(self, *_):
         assert any('percent' in msg and msg['percent'] == 100.0
                    for msg in notify.messages if msg)
 
+    @patch('tools.toolchains.arm.ARM_STD.parse_dependencies',
+           return_value=["foo"])
+    @patch('tools.toolchains.mbedToolchain.need_update',
+           side_effect=[i % 2 for i in range(3000)])
+    @patch('os.mkdir')
+    @patch('tools.toolchains.mbedToolchain.dump_build_profile')
+    @patch('tools.utils.run_cmd', return_value=(b'', b'', 0))
+    def test_compile_legacy_sources_always_complete_build(self, *_):
+        """Test that compile_legacy_sources() completes."""
+        notify = MockNotifier()
+        toolchain = prepare_toolchain(self.src_paths, self.build_path, self.target,
+                                      self.toolchain_name, notify=notify)
+
+        res = Resources(MockNotifier()).scan_with_toolchain(
+            self.src_paths, toolchain)
+
+        toolchain.RESPONSE_FILES=False
+        toolchain.config_processed = True
+        toolchain.config_file = "junk"
+        toolchain.compile_legacy_sources(res)
+
+        assert any('percent' in msg and msg['percent'] == 100.0
+                   for msg in notify.messages if msg)
+
+    def test_dirs_exclusion_from_file_to_compile(self, *_):
```

Comment:
This only seems to work when run from a directory where source files can be found, otherwise it fails (I have discovered that is also the reason the other tests fail when a run them).

A unit test which just tests the method could be written, e.g.

```suggestion
    def test_dirs_exclusion_from_file_to_compile(self):
        files_to_compile = [
            FileRef(name='platform/TARGET_CORTEX_M/TOOLCHAIN_ARM/except.S',
                    path='./platform/TARGET_CORTEX_M/TOOLCHAIN_ARM/except.S'),
            FileRef(
                name='rtos/TARGET_CORTEX/rtx5/RTX/Source/TOOLCHAIN_ARM/TARGET_RTOS_M4_M7/targets_irq_cm4f.S',
                path='./rtos/TARGET_CORTEX/rtx5/RTX/Source/TOOLCHAIN_ARM/TARGET_RTOS_M4_M7/targets_irq_cm4f.S')
        ]
        exclude_dirs = ['platform/', 'drivers/', 'targets/']
        expected_compilation_queue = [
            FileRef(
                name='rtos/TARGET_CORTEX/rtx5/RTX/Source/TOOLCHAIN_ARM/TARGET_RTOS_M4_M7/targets_irq_cm4f.S',
                path='./rtos/TARGET_CORTEX/rtx5/RTX/Source/TOOLCHAIN_ARM/TARGET_RTOS_M4_M7/targets_irq_cm4f.S')
        ]

        compilation_queue = mbedToolchain._exclude_files_from_build(files_to_compile, exclude_dirs)

        self.assertEquals(compilation_queue, expected_compilation_queue)
```

Reply:
@madchutney this is not for me to fix. All tests on this file are written like that. It can be fixed as part of a different PR. Let's not make this bigger than what it was meant to be please.

Reply:
Some extra changes were needed so I could not just use your suggested change.
Here is the change https://github.com/ARMmbed/mbed-os/pull/11114/commits/951b6e0e7ced16ef585eb21efee4fef0245373ca and I hope we can now close this chapter.